ompi/mca/coll/base/coll_base_allreduce.c

```diff
@@ -367,8 +367,8 @@ ompi_coll_base_allreduce_intra_ring(const void *sbuf, void *rbuf, int count,
         return MPI_SUCCESS;
     }
 
-    /* Special case for count less than size - use recursive doubling */
-    if (count < size) {
+    /* Special case for count less than size or non-commutative ops - use recursive doubling */
```

Comment:
My concern about this is that neither recursive_doubling or intra_ring are suitable algorithms for non-commutative operators.

Reply:
The documentation for recursive-doubling says that it works with commutative ops. I changed the checks again a bit and force-pushed.

Reply:
ok, it does indeed support non-commutative but associative ops (line 227 in the file).